### ISSUE TITLE
The sum of all intermediate losses, is slightly different than the sum post backward()

### ISSUE THREAD

**Issue by ProGamerGov** [pytorch/pytorch#6963]
## Issue description

When adding up all the intermediate losses, I get a slightly different value than the total loss.

## Code example

```
import torch
import torch.nn as nn
from torch.autograd import Variable
crit = nn.MSELoss()

####################
torch.manual_seed(876)
a = Variable(torch.randn(3,64,64), requires_grad=True)

torch.manual_seed(875)
b = Variable(torch.randn(3,64,64))
c = crit(a,b)
#####################
torch.manual_seed(874)
d = Variable(torch.randn(3,64,64), requires_grad=True)

torch.manual_seed(873)
e = Variable(torch.randn(3,64,64))
f = crit(d,e)

####################
torch.manual_seed(876)
h = Variable(torch.randn(3,64,64), requires_grad=True)

torch.manual_seed(875)
i = Variable(torch.randn(3,64,64))
j = crit(h,i)
#####################

test = (j, f, c)

g = 0
for loss in test: 
   g+=loss 
g.backward(retain_graph=True)


print("Rounded Results:")
print("Using += before backward(): " + str(float(g)))
print("Using a single float(): " + str( float(j + f + c)  ))
print("Using a single item(): " + str( (j + f + c).item()  ))

print("Working Methods:")
print("Adding j, f, c, separately using float(): " + str(float(j) + float(f) + float(c) ))
print("Adding j, f, c, separately using item(): " + str(j.item() + f.item() + c.item() ))

print("The difference Between Adding Methods:")
print("Difference: " + str( (float(j) + float(f) + float(c)) - float(g) ))
```

The above code will output: 

```
Rounded Results:
Using +=: 6.0663561821
Using a single float(): 6.0663561821
Using a single item(): 6.0663561821
Working Methods:
Adding j, f, c, separately using float(): 6.06635594368
Adding j, f, c, separately using item(): 6.06635594368
The difference Between Adding Methods:
Difference: -2.38418579102e-07
```

The loss variables `j`, `f`, and `c` are equal to: 

`j` = `2.03060102463`

`f` = `2.00515389442`

`c` = `2.03060102463`

When I add them together in python manually, I don't get the rounded result:

```
>>> a = 2.03060102463 + 2.00515389442 + 2.03060102463
>>> print(a)
6.06635594368
```

## System Info

- PyTorch version: 0.4.0
- Is debug build: No
- CUDA used to build PyTorch: 8.0.61

- OS: Ubuntu 16.04.3 LTS
- GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.5) 5.4.0 20160609
- CMake version: version 3.5.1

- Python version: 2.7
- Is CUDA available: Yes
- CUDA runtime version: Could not collect
- GPU models and configuration: GPU 0: Tesla K80
- Nvidia driver version: 384.90
- cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.0.3
/usr/lib/x86_64-linux-gnu/libcudnn_static_v7.a
/usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudnn.so
/usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudnn.so.7
/usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudnn.so.7.0.3
/usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudnn_static.a

- Versions of relevant libraries:
[pip] Could not collect
[conda] Could not collect

**Comment by zou3519** [comment]
I'm confused. 

>When I add them together in python manually, I don't get the rounded result:
>>> a = 2.03060102463 + 2.00515389442 + 2.03060102463
>>> print(a)
6.06635594368

This number appears equal to the other numbers you got:

>>> Adding j, f, c, separately using float(): 6.06635594368
>>> Adding j, f, c, separately using item(): 6.06635594368

Where is the discrepancy?

**Comment by ProGamerGov** [comment]
The discrepancy is between `g` and `j` + `f` + `c`. 

`g` is made up of `j` + `f` + `c`, so it should be equal to `j` + `f` + `c`. But instead it's slightly different.

**Comment by zou3519** [comment]
Gotcha, thank you for the clarification. I'm not sure if this is a bug or if it is just how floating point precision works. The difference you noted:

> Difference: -2.38418579102e-07

is normal for floating-point precision. Another way of summing gives the same result as `g`:

```
summed = torch.sum(torch.cat([j.unsqueeze(0), f.unsqueeze(0), c.unsqueeze(0)])).item()
(g - summed).item()
```

**Comment by ProGamerGov** [comment]
Looking in to the cause of the difference in my code, one these two methods of adding the intermediate loss for printing are used:


```
                totalLoss += float(mod.loss)
```

                totalLoss += mod.loss

If I use `+= float(mod.loss)`, I get a different result than just using `+= mod.loss`. I would imagine that using `+= mod.loss.item()` is the same as `+= float(mod.loss)`. 

Maybe I'm just using the wrong method of printing the losses? Which of the 3 should I be using?

**Comment by ngimel** [comment]
closing due to age.
